File: python/hashlock/ws.py
"""Maker feed over WebSocket (/v1/ws), using the ``websockets`` library (async).

Authenticate with an API key that has the ``maker`` scope, receive a snapshot of the public order book,
then a live stream of quotable RFQs; submit quotes on the same socket.
"""
from __future__ import annotations

import json
from typing import Any, AsyncIterator

try:
    import websockets  # type: ignore
except ImportError:  # pragma: no cover
    websockets = None  # the maker feed is optional; install `hashlock-sdk[ws]`

DEFAULT_WS = "wss://api-dev.hashlock.markets/v1/ws"


async def maker_feed(api_key: str, url: str = DEFAULT_WS) -> AsyncIterator[dict[str, Any]]:
    """Async-iterate feed messages: {'type': 'snapshot'|'ready'|'rfq'|'quoted'|'error', ...}.

    Example::

        async for msg in maker_feed(api_key):
            if msg["type"] == "rfq":
                ...  # decide whether to quote
    """
    if websockets is None:
        raise RuntimeError("install `hashlock-sdk[ws]` (the `websockets` package) to use the maker feed")
    async with websockets.connect(url) as ws:
        await ws.send(json.dumps({"apiKey": api_key}))
        async for raw in ws:
            try:
                yield json.loads(raw)
            except json.JSONDecodeError:
                continue


async def submit_quote(ws: Any, rfq_id: str, quote_amount: str) -> None:
    """Submit a quote on an open maker-feed socket."""
    await ws.send(json.dumps({"quote": {"rfqId": rfq_id, "quoteAmount": quote_amount}}))

File: typescript/src/ws.ts
// Maker feed over WebSocket (/v1/ws). Authenticate with an API key that has the `maker` scope, receive a
// snapshot of the public order book, then a live stream of quotable RFQs; submit quotes on the same socket.
// Uses the global WebSocket (Node 22+, browsers, edge). In older Node, pass one via `options.WebSocket`.

import type { Rfq } from './types.js';

export interface MakerFeedOptions {
  apiKey: string;
  /** WS URL. Default derives from the REST base: wss://api-dev.hashlock.markets/v1/ws */
  url?: string;
  WebSocket?: typeof WebSocket;
  onSnapshot?: (rfqs: Rfq[]) => void;
  onRfq?: (rfq: Rfq, kind: string) => void;
  onQuoted?: (msg: { rfqId: string; threadId: string }) => void;
  onError?: (message: string) => void;
  onClose?: () => void;
}

const DEFAULT_WS = 'wss://api-dev.hashlock.markets/v1/ws';

export class MakerFeed {
  private ws?: WebSocket;
  private readonly opts: MakerFeedOptions;
  constructor(opts: MakerFeedOptions) {
    this.opts = opts;
  }

  connect(): Promise<void> {
    const Ctor = this.opts.WebSocket ?? (globalThis as { WebSocket?: typeof WebSocket }).WebSocket;
    if (!Ctor) throw new Error('MakerFeed: no WebSocket available — pass options.WebSocket');
    const ws = new Ctor(this.opts.url ?? DEFAULT_WS);
    this.ws = ws;
    return new Promise((resolve, reject) => {
      ws.onopen = () => ws.send(JSON.stringify({ apiKey: this.opts.apiKey }));
      ws.onerror = () => { this.opts.onError?.('websocket error'); reject(new Error('websocket error')); };
      ws.onclose = () => this.opts.onClose?.();
      ws.onmessage = (ev: MessageEvent) => {
        let m: { type?: string; rfqs?: Rfq[]; rfq?: Rfq; kind?: string; rfqId?: string; threadId?: string; error?: string };
        try { m = JSON.parse(String(ev.data)); } catch { return; }
        switch (m.type) {
          case 'snapshot': this.opts.onSnapshot?.(m.rfqs ?? []); break;
          case 'ready': resolve(); break;
          case 'rfq': if (m.rfq) this.opts.onRfq?.(m.rfq, m.kind ?? 'created'); break;
          case 'quoted': this.opts.onQuoted?.({ rfqId: m.rfqId!, threadId: m.threadId! }); break;
          case 'error': this.opts.onError?.(m.error ?? 'error'); break;
        }
      };
    });
  }

  /** Submit a quote for an RFQ over the socket. */
  quote(rfqId: string, quoteAmount: string): void {
    this.ws?.send(JSON.stringify({ quote: { rfqId, quoteAmount } }));
  }

  close(): void {
    this.ws?.close();
  }
}
